File: app/src/main/java/org/mifos/mobilebanking/utils/DateHelper.java
/*
 * This project is licensed under the open source MPL V2.
 * See https://github.com/openMF/android-client/blob/master/LICENSE.md
 */

package org.mifos.mobilebanking.utils;

import android.util.Log;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;

/**
 * Created by ishankhanna on 30/05/14.
 * <p>This is a helper class that will be used to convert List<Interger> Type Dates
 * from MifosX into Simple Strings or Date Formats</p>
 */
public class DateHelper {

    public static final String LOG_TAG = DateHelper.class.getSimpleName();

    public static final String FORMAT_dd_MMMM_yyyy = "dd MMMM yyyy";

    public static List<Integer> getCurrentDate(String dateFormat, String separator) {
        List<Integer> date = new ArrayList<>();

        String s = new SimpleDateFormat(dateFormat, Locale.getDefault()).format(new Date());
        for (String str : s.split(separator)) {
            date.add(Integer.parseInt(str));
        }

        return date;
    }

    /**
     * the result string uses the list given in a reverse order ([x, y, z] results in "z y x")
     *
     * @param integersOfDate [year-month-day] (ex [2016, 4, 14])
     * @return date in the format day month year (ex 14 Apr 2016)
     */
    public static String getDateAsString(List<Integer> integersOfDate) {

        StringBuilder stringBuilder = new StringBuilder();
        if (integersOfDate != null) {
            stringBuilder.append(integersOfDate.get(2))
                    .append(' ')
                    .append(getMonthName(integersOfDate.get(1)))
                    .append(' ')
                    .append(integersOfDate.get(0));
        }
        return stringBuilder.toString();
    }

    public static String getDateAsString(List<Integer> integersOfDate, String pattern) {
        return DateHelper.getFormatConverter("dd MMM yyyy",
                pattern, DateHelper.getDateAsString(integersOfDate));

    }

    /**
     * This Method converting the dd-MM-yyyy format type date string into dd MMMM yyyy
     *
     * @param format     Final Format of date string
     * @param dateString date string
     * @return dd MMMM yyyy format date string.
     */
    public static String getSpecificFormat(String format, String dateString) {
        SimpleDateFormat pickerFormat = new SimpleDateFormat("dd-MM-yyyy", Locale.ENGLISH);
        SimpleDateFormat finalFormat = new SimpleDateFormat(format, Locale.ENGLISH);
        Date date = null;
        try {
            date = pickerFormat.parse(dateString);
        } catch (ParseException e) {
            Log.d(LOG_TAG, e.getLocalizedMessage());
        }
        return finalFormat.format(date);
    }

    public static String getFormatConverter(String currentFormat, String requiredFormat,
            String dateString) {
        SimpleDateFormat pickerFormat = new SimpleDateFormat(currentFormat, Locale.ENGLISH);
        SimpleDateFormat finalFormat = new SimpleDateFormat(requiredFormat, Locale.ENGLISH);
        Date date = null;
        try {
            date = pickerFormat.parse(dateString);
        } catch (ParseException e) {
            Log.d(LOG_TAG, e.getLocalizedMessage());
        }
        return finalFormat.format(date);
    }

    /**
     * @param month an integer from 1 to 12
     * @return string representation of the month like Jan or Feb..etc
     */
    public static String getMonthName(int month) {
        String monthName = "";
        switch (month) {
            case 1:
                monthName = "Jan";
                break;
            case 2:
                monthName = "Feb";
                break;
            case 3:
                monthName = "Mar";
                break;
            case 4:
                monthName = "Apr";
                break;
            case 5:
                monthName = "May";
                break;
            case 6:
                monthName = "Jun";
                break;
            case 7:
                monthName = "Jul";
                break;
            case 8:
                monthName = "Aug";
                break;
            case 9:
                monthName = "Sep";
                break;
            case 10:
                monthName = "Oct";
                break;
            case 11:
                monthName = "Nov";
                break;
            case 12:
                monthName = "Dec";
                break;
        }
        return monthName;
    }

    public static long getDateAsLongFromString(String dateStr, String pattern) {
        SimpleDateFormat sdf = new SimpleDateFormat(pattern);
        Date date = null;

        try {
            date = sdf.parse(dateStr);
        } catch (ParseException e) {
            Log.d("TAG", e.getMessage());
        }
        return date.getTime();
    }


    public static long getDateAsLongFromList(List<Integer> integersOfDate) {
        String dateStr = getDateAsString(integersOfDate);
        return getDateAsLongFromString(dateStr, "dd MMM yyyy");
    }

    public static long subtractWeeks(int number) {
        Calendar calendar = Calendar.getInstance();
        calendar.add(Calendar.WEEK_OF_YEAR, -number);
        return calendar.getTimeInMillis();
    }


    public static long subtractMonths(int number) {
        Calendar calendar = Calendar.getInstance();
        calendar.add(Calendar.MONTH, -number);
        return calendar.getTimeInMillis();
    }

    public static String getDateAsStringFromLong(long timeInMillis) {
        SimpleDateFormat sdf = new SimpleDateFormat("dd MMM yyyy");
        return sdf.format(new Date(timeInMillis));
    }

    public static String getDateAndTimeAsStringFromLong(long timeInMillis) {
        SimpleDateFormat sdf = new SimpleDateFormat("HH:mm a dd MMM yyyy");
        return sdf.format(new Date(timeInMillis));
    }

}

File: core/common/src/main/java/org/mifos/mobile/core/common/utils/DateHelper.kt
/*
 * Copyright 2024 Mifos Initiative
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * See https://github.com/openMF/mobile-mobile/blob/master/LICENSE.md
 */
package org.mifos.mobile.core.common.utils

import android.util.Log
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

object DateHelper {

    private val LOG_TAG: String? = DateHelper::class.java.simpleName
    const val FORMAT_MMMM = "dd MMMM yyyy"
    const val FORMAT_MM = "dd-MM-yyyy"
    fun getCurrentDate(dateFormat: String?, separator: String): List<Int> {
        val date: MutableList<Int> = ArrayList()
        val s = SimpleDateFormat(dateFormat, Locale.getDefault()).format(Date())
        for (str in s.split(separator.toRegex()).toTypedArray()) {
            date.add(str.toInt())
        }
        return date
    }

    /**
     * the result string uses the list given in a reverse order ([x, y, z] results in "z y x")
     *
     * @param integersOfDate [year-month-day] (ex [2016, 4, 14])
     * @return date in the format day month year (ex 14 Apr 2016)
     */
    fun getDateAsString(integersOfDate: List<Int?>?): String {
        val stringBuilder = StringBuilder()
        if (integersOfDate != null && integersOfDate.size >= 3) {
            stringBuilder.append(integersOfDate[2])
                .append(' ')
                .append(getMonthName(integersOfDate[1]))
                .append(' ')
                .append(integersOfDate[0])
        }
        return stringBuilder.toString()
    }

    fun getDateAsString(integersOfDate: List<Int>?, pattern: String?): String {
        return getFormatConverter(
            "dd MMM yyyy",
            pattern,
            getDateAsString(integersOfDate),
        )
    }

    /**
     * This Method converting the dd-MM-yyyy format type date string into dd MMMM yyyy
     *
     * @param format     Final Format of date string
     * @param dateString date string
     * @return dd MMMM yyyy format date string.
     */
    fun getSpecificFormat(format: String?, dateString: String?): String {
        val pickerFormat = SimpleDateFormat("dd-MM-yyyy", Locale.ENGLISH)
        val finalFormat = SimpleDateFormat(format, Locale.ENGLISH)
        var date: Date? = null
        try {
            date = pickerFormat.parse(dateString)
        } catch (e: ParseException) {
            Log.d(LOG_TAG, e.localizedMessage)
        }
        return finalFormat.format(date)
    }

    fun getSpecificFormat(format: String?, dateLong: Long?): String? {
        try {
            return SimpleDateFormat(format, Locale.getDefault()).format(dateLong)
        } catch (e: ParseException) {
            Log.d(LOG_TAG, e.localizedMessage)
            return null
        }
    }

    fun getFormatConverter(
        currentFormat: String?,
        requiredFormat: String?,
        dateString: String?,
    ): String {
        val pickerFormat = SimpleDateFormat(currentFormat, Locale.ENGLISH)
        val finalFormat = SimpleDateFormat(requiredFormat, Locale.ENGLISH)
        var date: Date? = null
        try {
            date = pickerFormat.parse(dateString)
        } catch (e: ParseException) {
            Log.d(LOG_TAG, e.localizedMessage)
        }
        return finalFormat.format(date)
    }

    /**
     * @param month an integer from 1 to 12
     * @return string representation of the month like Jan or Feb..etc
     */
    private fun getMonthName(month: Int?): String {
        var monthName = ""
        when (month) {
            1 -> monthName = "Jan"
            2 -> monthName = "Feb"
            3 -> monthName = "Mar"
            4 -> monthName = "Apr"
            5 -> monthName = "May"
            6 -> monthName = "Jun"
            7 -> monthName = "Jul"
            8 -> monthName = "Aug"
            9 -> monthName = "Sep"
            10 -> monthName = "Oct"
            11 -> monthName = "Nov"
            12 -> monthName = "Dec"
        }
        return monthName
    }

    fun getDateAsLongFromString(dateStr: String?, pattern: String?): Long? {
        val sdf = SimpleDateFormat(pattern, Locale.getDefault())
        var date: Date? = null
        try {
            date = sdf.parse(dateStr)
        } catch (e: ParseException) {
            Log.d("TAG", e.message ?: "")
        }
        return date?.time
    }

    @JvmStatic
    fun getDateAsLongFromList(integersOfDate: List<Int>?): Long? {
        val dateStr = getDateAsString(integersOfDate)
        return getDateAsLongFromString(dateStr, "dd MMM yyyy")
    }

    fun subtractWeeks(number: Int): Long {
        val calendar = Calendar.getInstance()
        calendar.add(Calendar.WEEK_OF_YEAR, -number)
        return calendar.timeInMillis
    }

    fun subtractMonths(number: Int): Long {
        val calendar = Calendar.getInstance()
        calendar.add(Calendar.MONTH, -number)
        return calendar.timeInMillis
    }

    fun getDateAsStringFromLong(timeInMillis: Long?): String {
        val sdf = SimpleDateFormat("dd MMM yyyy", Locale.getDefault())
        return sdf.format(timeInMillis?.let { Date(it) })
    }

    @JvmStatic
    fun getDateAndTimeAsStringFromLong(timeInMillis: Long?): String {
        val sdf = SimpleDateFormat("HH:mm a dd MMM yyyy", Locale.getDefault())
        return sdf.format(timeInMillis?.let { Date(it) })
    }
}

fun getTodayFormatted(): String =
    SimpleDateFormat("dd-MM-yyyy", Locale.getDefault()).format(Calendar.getInstance().time)
